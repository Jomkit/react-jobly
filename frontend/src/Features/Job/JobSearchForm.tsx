import React, { useState } from 'react';
import JoblyApi from '../../api.ts';
import { Col, FormGroup, Input, Label, Row } from 'reactstrap'
import { jobsFormInterface } from '../../types.ts';

const JobSearchForm = ({ setData }: { setData: Function }) => {
    const initialState: jobsFormInterface = {
        title: "",
        minSalary: 0,
        hasEquity: false
    }
    const [ formData, setFormData ] = useState(initialState);

    // Need to apply filter to form data
    async function getFiltered(filterParams: jobsFormInterface) {
        try{
            const res = await JoblyApi.getJobs(filterParams);
            setData(res);
        } catch(e: any){
            console.log("Oops, something went wrong...");
            console.error(e);
        }
    }
    
    function handleChange(evt: React.ChangeEvent<HTMLInputElement>){
        const { name, value } = evt.target;
        setFormData((data: jobsFormInterface) => ({
            ...data,
            [name]: value,
        }));
    }

    function handleSubmit(evt: React.FormEvent){
        evt.preventDefault();
        getFiltered(formData);
    }

  return (
    <form className='bg-white mx-5 p-1 border border-3 border-black rounded' onSubmit={handleSubmit}>
        <Row className='m-lg-4 align-items-center'>
            <Col lg={6}>
                <FormGroup>
                    <Label
                        className='visually-hidden'
                        for='title'
                    >
                        Job Title
                    </Label>
                    <Input
                    id='title'
                    name='title'
                    placeholder='search job titles'
                    type="text"
                    value={formData.title}
                    onChange={handleChange}
                    />            
                </FormGroup>
            </Col>
            <Col lg={2}>
                <FormGroup floating>
                    <Input 
                        id="minSalary"
                        name="minSalary"
                        placeholder='0'
                        min={0}
                        type="number"
                        bsSize='sm'
                        value={formData.minSalary}
                        onChange={handleChange}
                    />
                    <Label for="minSalary">Minimum Salary</Label>
                </FormGroup>
            </Col>
            <Col lg={2}>
                <FormGroup check inline>
                    <Input 
                        id="hasEquity"
                        name="hasEquity"
                        type="checkbox"
                        value="true"
                        onChange={handleChange}
                    />
                    <Label check for="hasEquity">Has Equity?</Label>
                </FormGroup>
            </Col>

            <Col className='col-2'>
                <button className='btn bg-primary text-light btn-block'>Search</button>
            </Col>
        </Row>
    </form>
  )
}

export default JobSearchForm