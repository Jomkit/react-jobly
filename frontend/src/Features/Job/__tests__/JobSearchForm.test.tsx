import { fireEvent, render } from "@testing-library/react"
import JobSearchForm from "../JobSearchForm"
import JoblyApi from "../../../api.ts";
import { MemoryRouter } from "react-router-dom";

// Mock api
vi.mock("../../../api.ts");
JoblyApi.getJobs = vi.fn();
const mockSetData = vi.fn();

const TestJobSearchForm = () => {
    return (
    <MemoryRouter>
        <JobSearchForm setData={mockSetData} />
    </MemoryRouter>
    )
}

describe("Job Search Form", () => {
    it('should render without crashing', () => { 
        render(<TestJobSearchForm />);
    });

    it('should match snapshot', () => { 
        const { asFragment } = render(<TestJobSearchForm />);
        expect(asFragment()).toMatchSnapshot();
    });

    it('should call JoblyApi.getJobs when search submit button pressed', () => { 
        const { getByText, getByLabelText } = render(<TestJobSearchForm />);

        const submitBtn = getByText("Search");
        const nameInput = getByLabelText("Job Title");
        fireEvent.change(nameInput, { target: { value: "test job 1" } });
        fireEvent.click(submitBtn);

        expect(JoblyApi.getJobs).toHaveBeenCalled();
        expect(JoblyApi.getJobs).toHaveBeenCalledWith({
            title: "test job 1", 
            minSalary: 0,
            hasEquity: false
        });
    })
})