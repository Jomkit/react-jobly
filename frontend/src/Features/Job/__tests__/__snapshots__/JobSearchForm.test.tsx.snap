// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Job Search Form > should match snapshot 1`] = `
<DocumentFragment>
  <form
    class="bg-white mx-5 p-1 border border-3 border-black rounded"
  >
    <div
      class="m-lg-4 align-items-center row"
    >
      <div
        class="col-lg-6"
      >
        <div
          class="mb-3"
        >
          <label
            class="visually-hidden form-label"
            for="title"
          >
            Job Title
          </label>
          <input
            class="form-control"
            id="title"
            name="title"
            placeholder="search job titles"
            type="text"
            value=""
          />
        </div>
      </div>
      <div
        class="col-lg-2"
      >
        <div
          class="mb-3 form-floating"
        >
          <input
            class="form-control-sm form-control"
            id="minSalary"
            min="0"
            name="minSalary"
            placeholder="0"
            type="number"
            value="0"
          />
          <label
            class="form-label"
            for="minSalary"
          >
            Minimum Salary
          </label>
        </div>
      </div>
      <div
        class="col-lg-2"
      >
        <div
          class="form-check form-check-inline"
        >
          <input
            class="form-check-input"
            id="hasEquity"
            name="hasEquity"
            type="checkbox"
            value="true"
          />
          <label
            class="form-check-label"
            for="hasEquity"
          >
            Has Equity?
          </label>
        </div>
      </div>
      <div
        class="col-2 col"
      >
        <button
          class="btn bg-primary text-light btn-block"
        >
          Search
        </button>
      </div>
    </div>
  </form>
</DocumentFragment>
`;
